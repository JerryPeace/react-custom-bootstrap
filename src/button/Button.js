import PropTypes from 'prop-types';
import React from 'react';
import classNames from 'classnames';
import styleMaps from '../styleMaps';
import CustomPropTypes from '../utils/CustomPropTypes';
import bootstrapUtils from '../utils/bootstrapUtils';
import faUtils from '../utils/faUtils';
import Loader from '../loader/Loader';
import TmIcon from '../icon/TmIcon';

class Button extends React.Component {

    static propTypes = {
        isLoading: PropTypes.bool,
        active: PropTypes.bool,
        disabled: PropTypes.bool,
        block: PropTypes.bool,
        // navItem: React.PropTypes.bool,
        // navDropdown: React.PropTypes.bool,
        /**
         * You can use a custom element for this component
         */
        componentClass: CustomPropTypes.elementType,
        href: PropTypes.string,
        target: PropTypes.string,
        onClick: PropTypes.func,
        type: PropTypes.string,
        buttonText: PropTypes.oneOfType([PropTypes.string, PropTypes.element]),
        isDropdown: PropTypes.bool,
        isCaretStyle: PropTypes.bool,
        faStyle: PropTypes.string,
        faProps: PropTypes.object,
        tmStyle: PropTypes.string,
        tmProps: PropTypes.object,
        iconEl: PropTypes.node,
        bsClass: PropTypes.string,
        bsStyle: PropTypes.string,
        bsSize: PropTypes.oneOf(['lg', 'sm', 'xs']),
        isIconOnly: PropTypes.bool,
        standalone: PropTypes.bool
    };

    static defaultProps = {
        isLoading: false,
        active: false,
        block: false,
        bsClass: 'button',
        bsStyle: 'default',
        disabled: false,
        standalone: true,
        type: 'button',
        isDropdown: false,
        isCaretStyle: false,
        faProps: {}
    };

    getComponent() {
        const {componentClass, href, target} = this.props;
        if (componentClass) {
            return componentClass;
        } else {
            return href || target ? 'a' : 'button';
        }
    }

    getClassName() {
        const {'dropdown-toggle': dropdownToggle} = styleMaps.CLASSES;
        const bsCls = bootstrapUtils.getCls(this.props);

        const {
            className,
            active: isActive,
            block: isBlock,
            disabled: isDisabled,
            isDropdown,
            isIconOnly
        } = this.props;

        const elementCls = {
            active: isActive,
            'btn-block': isBlock,
            disabled: isDisabled,
            isDropdown,
            'btn-icon-only': isIconOnly,
            [dropdownToggle]: isDropdown
        };
        return classNames(bsCls, elementCls, className);
    }

    renderIcon() {
        const {
            isLoading,
            iconEl,
            faStyle,
            faProps,
            tmStyle,
            tmProps
        } = this.props;
        const faClass = faUtils.getCls(this.props);

        let icon = null;

        if (isLoading) {
            icon = <Loader bsSize={styleMaps.SIZES.sm} />;
        } else if (iconEl) {
            icon = iconEl;
        } else if (tmStyle) {
            icon = <TmIcon name={tmStyle} size={16} { ...tmProps } />
        } else if (faClass) {
            icon = <i {...faProps} className={faClass} />;
        }

        if (icon) {
            return (
                <span className={styleMaps.CLASSES['btn-icon']}>
                    {icon}
                </span>
            );
        } else {
            return null;
        }
    }

    renderContent() {
        const {buttonText, children} = this.props;
        if (buttonText || children) {
            return (
                <span className={styleMaps.CLASSES['btn-content']} >
                    {buttonText}
                    {children}
                </span>
            );
        } else {
            return null;
        }
    }

    renderCaret() {
        const {caret} = styleMaps.CLASSES;
        if (this.props.isCaretStyle) {
            return <span className={caret} />;
        } else {
            return null;
        }
    }


    render() {
        const Component = this.getComponent();
        let {isLoading, iconEl, active, block, bsClass, bsStyle, bsSize, standalone, buttonText, isDropdown, isCaretStyle, tmStyle, tmProps, faStyle, faProps, isIconOnly, componentClass, ...rest} = this.props;

        if (Component === 'a') {
            delete rest.disabled;
            rest.href = this.props.href || '#';
        }

        return (
            <Component
                key={'button'}
                {...rest}
                className={this.getClassName()} >
                {this.renderIcon()}
                {this.renderContent()}
                {this.renderCaret()}
            </Component>
        );
    }
}

export default Button;
