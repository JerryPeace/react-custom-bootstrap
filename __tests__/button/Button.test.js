import React from 'react';
import {mount} from 'enzyme';
import Button from 'button/Button';

describe('Button', () => {
    let button;

    test('without children or buttonText', () => {
        button = mount(
            <Button></Button>
        );
        expect(button).toMatchSnapshot();
    });

    test('with button label', () => {
        button = mount(
            <Button btnLabel='Test Label' >
                children
            </Button>
        );
        expect(button).toMatchSnapshot();
    });

    test('with customize icon in <a> wrapper', () => {
        button = mount(
            <Button
                buttonText='Click'
                iconEl='<span>icon</span>'
                target='_blank'
            >
                children
            </Button>
        );
        expect(button).toMatchSnapshot();
    });

    test('with loading status in <div> wrapper', () => {
        button = mount(
            <Button
                componentClass='div'
                buttonText='Click'
                isLoading
            >
            </Button>
        );
        expect(button).toMatchSnapshot();
    });

    test('with fa icon and caret style in <a> wrapper', () => {
        button = mount(
            <Button
                buttonText='Click'
                href='test.com'
                faStyle='plus'
                isCaretStyle
            >
            </Button>
        );
        expect(button).toMatchSnapshot();
    });
});
