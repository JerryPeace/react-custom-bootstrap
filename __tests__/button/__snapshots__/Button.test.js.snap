// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button with button label 1`] = `
<Button
  active={false}
  block={false}
  bsClass="button"
  bsStyle="default"
  btnLabel="Test Label"
  disabled={false}
  faProps={Object {}}
  isCaretStyle={false}
  isDropdown={false}
  isLoading={false}
  standalone={true}
  type="button"
>
  <button
    btnLabel="Test Label"
    className="btn btn-default"
    disabled={false}
    type="button"
  >
    <span
      className="btn-content"
    >
      children
    </span>
  </button>
</Button>
`;

exports[`Button with customize icon in <a> wrapper 1`] = `
<Button
  active={false}
  block={false}
  bsClass="button"
  bsStyle="default"
  buttonText="Click"
  disabled={false}
  faProps={Object {}}
  iconEl="<span>icon</span>"
  isCaretStyle={false}
  isDropdown={false}
  isLoading={false}
  standalone={true}
  target="_blank"
  type="button"
>
  <a
    className="btn btn-default"
    href="#"
    target="_blank"
    type="button"
  >
    <span
      className="btn-icon"
    >
      &lt;span&gt;icon&lt;/span&gt;
    </span>
    <span
      className="btn-content"
    >
      Click
      children
    </span>
  </a>
</Button>
`;

exports[`Button with fa icon and caret style in <a> wrapper 1`] = `
<Button
  active={false}
  block={false}
  bsClass="button"
  bsStyle="default"
  buttonText="Click"
  disabled={false}
  faProps={Object {}}
  faStyle="plus"
  href="test.com"
  isCaretStyle={true}
  isDropdown={false}
  isLoading={false}
  standalone={true}
  type="button"
>
  <a
    className="btn btn-default"
    href="test.com"
    type="button"
  >
    <span
      className="btn-icon"
    >
      <i
        className="fa fa-plus"
      />
    </span>
    <span
      className="btn-content"
    >
      Click
    </span>
    <span
      className="caret"
    />
  </a>
</Button>
`;

exports[`Button with loading status in <div> wrapper 1`] = `
<Button
  active={false}
  block={false}
  bsClass="button"
  bsStyle="default"
  buttonText="Click"
  componentClass="div"
  disabled={false}
  faProps={Object {}}
  isCaretStyle={false}
  isDropdown={false}
  isLoading={true}
  standalone={true}
  type="button"
>
  <div
    className="btn btn-default"
    disabled={false}
    type="button"
  >
    <span
      className="btn-icon"
    >
      <Loader
        bsClass="tm-loader"
        bsSize="sm"
      >
        <span
          className="loader"
        >
          <span
            className="tm-loader tm-loader-sm"
          />
        </span>
      </Loader>
    </span>
    <span
      className="btn-content"
    >
      Click
    </span>
  </div>
</Button>
`;

exports[`Button without children or buttonText 1`] = `
<Button
  active={false}
  block={false}
  bsClass="button"
  bsStyle="default"
  disabled={false}
  faProps={Object {}}
  isCaretStyle={false}
  isDropdown={false}
  isLoading={false}
  standalone={true}
  type="button"
>
  <button
    className="btn btn-default"
    disabled={false}
    type="button"
  />
</Button>
`;
